src/firewall/server/decorators.py

```diff
@@ -137,7 +140,9 @@ def __call__(self, func):
         )
         @functools.wraps(func)
         def _impl(*args, **kwargs):
-            pass
+            GLib.idle_add(
+                run_dispatcher, self.dbus_interface, func.__name__, (args[1:])
```

Comment:
run_dispatcher already spawns the process in the background. It's not clear why this needs to happen on an idle handler. The call does not block, so it can happen right away.

Spawning it in the background, also means that all processes run in parallel and their order is non-deterministic. `GLib.idle_add()` does not help with that. We could help with that, by `run_dispatcher()` building a list of jobs, and only spawning one job at a time. But `GLib.idle_add()` doesn't achieve that.

Reply:
ok, maybe the advantage is to emit the D-Bus signal faster.

spawning a process, even if it's async, may still be somewhat slow...

Reply:
> The call does not block, so it can happen right away.

It does not block, but still takes time to fork+exec and wait on `systemd-run` to start the real process.

> Spawning it in the background, also means that all processes run in parallel and their order is non-deterministic

Right, this is noted in the man page. To be order deterministic firewalld would have to wait for each one to run to completion before starting the next. Could use `systemd-run --wait` and `Popen.poll()`. We would just keep adding the work to the mainloop with `idle_add()`.

That means building the list as you say, then calling each one from the mainloop, polling for completion, then moving on to the next dispatcher. Definitely more work than fire and forget. It does mean firewalld could report the exit status of the dispatcher.

If we use `systemd-run --wait` then we could instead directly call the dispatchers from firewalld and avoid the dependency on `systemd-run`. I was only using `systemd-run` because we are doing "fire and forget".

If we think that it's important to have in-order and non-parallel execution of dispatchers, then I'll make this change. I see the argument either way.